third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1184 @@
+                        .map(move |node_id| (env.get_node_loc(*node_id), msg.clone()))
+                })
+                .collect();
+            let msg = format!(
+                "recursion during function inlining not allowed: {} -> {}",
+                path_string,
+                func_env.get_name_str()
+            );
+            let loc = call_details.first_mut().unwrap().0.clone();
+            env.diag_with_labels(Severity::Error, &loc, &msg, call_details);
+        }
+        return Err(());
+    }
+
+    // Compute post-order of inlined_functions which call others.
+    let po_inline_functions = postorder(
+        &inline_functions_calling_others,
+        &inline_functions_with_callees,
+    );
+
+    // Identify subset of non-inline functions which call inline functions.  Order doesn't matter here.
+    let non_inline_functions_needing_inlining: Vec<QualifiedFunId> = function_callees
+        .iter()
+        .filter(|(fnid, callees)| {
+            !env.get_function(**fnid).is_inline()
+                && callees
+                    .iter()
+                    .any(|fnid2| env.get_function(*fnid2).is_inline())
+        })
+        .map(|(fnid, _callees)| fnid)
+        .cloned()
+        .collect();
+
+    let result: Vec<QualifiedFunId> =
+        chain(po_inline_functions, non_inline_functions_needing_inlining).collect();
+    Ok(result)
+}
+
+// Calculate a bottom-up traversal for entries, given the provided callee map.
+fn postorder<T: Ord + Copy + Debug>(
+    entries: &Vec<T>,
+    callee_map: &BTreeMap<T, BTreeSet<T>>,
+) -> Vec<T> {
+    let mut stack = Vec::new();
+    let mut visited = BTreeSet::new();
+    let mut grey = BTreeSet::new();
+    let mut postorder_num_to_node = Vec::new();
+    let mut node_to_postorder_num = BTreeMap::new();
+
+    for entry in entries {
+        if !visited.contains(&entry) {
+            visited.insert(entry);
+            stack.push(entry);
+            while let Some(curr) = stack.pop() {
+                if grey.contains(&curr) {
+                    let curr_num = postorder_num_to_node.len();
+                    postorder_num_to_node.push(*curr);
+                    node_to_postorder_num.insert(curr, curr_num);
+                } else {
+                    grey.insert(curr);
+                    stack.push(curr);
+                    if let Some(children) = callee_map.get(curr) {
+                        for child in children {
+                            if !visited.contains(child) {
+                                visited.insert(child);
+                                stack.push(child);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    postorder_num_to_node
+}
+
+// Check for cycles in a callee_map.
+// If there is a cycle, return one cyclical path.
+fn check_for_cycles<T: Ord + Copy + Debug>(
+    callee_map: &BTreeMap<T, BTreeSet<T>>,
+) -> BTreeSet<Vec<T>> {
```

Comment:
I think if there are multiple independent cycles it would be worthwhile reporting them all.  Detecting nonoverlapping cycles is a bit of algorithmic work I didn't feel up to at the moment, but I think it might be valuable to add in the future.